### DISCUSSION TITLE
Hooks/extension for AconfSave

### DISCUSSION THREAD

**Discussion by VorpalBlade** [CyberShadow/aconfmgr#153] [Ideas]
I would like some sort of hooking/extension system for `aconfmgr save`:

I use some helper functions (such as the passwd/group helpers from the wiki and my own systemd helpers that I also put there). I would like for `save` to use those instead of emitting `CreateFile /etc/passwd`,`CreateLink` etc (where appropriate).

The best way seems to be if there was a hooking/extension system of some sort to do this. Perhaps an array of filter functions that could either return false (return code != 0) to process as normal/continue to next filter or return true (0) and some output to use instead.

My interest here is for handling config files. While there could also be hooks for handling installed packages I don't see any obvious use cases.

**Comment by CyberShadow** [comment]
I'm open to discussing a design for this.

Ideally, such an extension mechanism could allow implementing plugins such as [user management](https://github.com/CyberShadow/aconfmgr/wiki/User-management).

> Perhaps an array of filter functions that could either return false (return code != 0) to process as normal/continue to next filter or return true (0) and some output to use instead.

Unfortunately we can't use this exact mechanism, as there's no way to check the exit code of a function in bash without altogether disabling error handling for the duration of the function's execution. It would either have to be a subprocess (not a function or fork/subshell) or communicate the result in some other way, like a string on stdout or setting a global variable.

**Reply by VorpalBlade** [reply]
I'm far more used to writing zsh than bash at this point, so I would leave the design in your hands.

I believe that either the filter should handle everything (emitting arbitrary strings) or let aconfmgr handle everything. It will of course be up to the filter to ensure that this results in an idempotent result. This should also work with user management, especially if it is modified to always sort the output (preferably the same way `grpck -s` and `pwck -s` sorts). This makes it non-sensitive to the order of `AddUser` calls. I have already made this sorting modification locally, but it needs to be cleaned up before it can be contributed back. As I will be away for a while soon this may take a bit.
